src/Unsafe/Linear.hs

```diff
@@ -20,42 +27,188 @@
+-- @
+toLinearL :: forall (l :: [Change]) f g m. ToLinearL l f g => f %m-> g
+toLinearL f = case unsafeLinearityProofL @l @f @g of
+  UnsafeRefl -> f
+
+-- | @ToLinearL l f g@ means that @f@ and @g@ are the same with the
+-- possible exception of the multiplicities of some of the first arrows,
+-- as described by the list @l@.
+type ToLinearL :: forall {rep :: RuntimeRep}. [Change] -> TYPE rep -> TYPE rep -> Constraint
+class ToLinearL l f g where
+  -- | @unsafeLinearityProofL \@l \@f \@g @ is a fake proof that @f@ and @g@ are
+  -- the same with the possible exception of the multiplicities of the first
+  -- arrows given by the list @l@.
+  unsafeLinearityProofL :: UnsafeEquality f g
+instance ( ToLinearN' ni f g, UnifyL ni f g, UnifyL ni g f
+         , ni ~ Length l, UseChanges l f g) => ToLinearL l f g where
+  unsafeLinearityProofL = prf @ni
+
+-- | Convert a GHC 'Nat' to a real inductive natural number.
+-- We use this because GHC 'Nat' offers a friendly API but
+-- it's a terrible pain for implementation.
+type ToINat :: Nat -> INat
+type family ToINat n where
+  ToINat 0 = 'Z
+  ToINat n = 'S (ToINat (n - 1))
+
+-- | The actual implementation of 'ToLinearN', using the inductive natural
+-- number it's handed.
+type ToLinearN' :: forall {rep :: RuntimeRep}. INat -> TYPE rep -> TYPE rep -> Constraint
+class ToLinearN' arrs f g where
+  prf :: UnsafeEquality f g
+instance ToLinearN' k fb gb => ToLinearN' ('S k) ((a :: TYPE repa) %p-> (fb :: TYPE repb)) (a %q-> (gb :: TYPE repb)) where
+  prf = case prf @k @fb @gb of
+    UnsafeRefl -> case unsafeEqualityProof @p @q of
+      UnsafeRefl -> UnsafeRefl
+instance ToLinearN' 'Z (a :: TYPE rep) (a :: TYPE rep) where
+  prf = UnsafeRefl
+
+-- | Plain old inductive natural numbers.
+data INat = Z | S INat
+
+-- | @UnifyL n f g@ walks @f@ to produce a constraint on @g@ ensuring that it
+-- is the same as @f@ except for the multiplicities of the first @n@ arrows.
+-- Using this symmetrically allows 'toLinearN' to infer large portions of
+-- either the \"from\" type or the \"to\" type from the other.
+type UnifyL :: forall rep. INat -> TYPE rep -> TYPE rep -> Constraint
+type family UnifyL arrs f g where
+  -- We need extra type signatures due to GHC ticket #17536.
+  UnifyL 'Z (a :: TYPE rep) b = a ~ b
+  UnifyL ('S n) ((a :: TYPE repa) %m-> (b :: TYPE repb)) g =
+    ( UnifyL n b (GetArrowRes g :: TYPE repb)
+    , g ~ (a %(GetArrowMult g)-> (GetArrowRes g :: TYPE repb)))
+
+  -- This is a bit unsatisfying. Can we use the fact that the count is positive
+  -- to *infer* that there's another arrow, squeezing out one last bit of
+  -- inference? The representation polymorphism seems to make this a bit
+  -- tricky, but maybe it's possible. Whether it's worth the trouble is another
+  -- question. There's probably not a huge practical benefit.
+  UnifyL ('S _) (_ :: TYPE rep) g =
+    TypeError ('Text "unsafeLinearityProofN: not enough arrows to satisfy request.")
```

Comment:
I don't think that `UnifyL` is necessary. You can simply phrase your `S` instance of `ToLinearN'` to force unification.

Something like
```haskell
 instance (ToLinearN' k fb gb, x~((a :: TYPE repa) %p-> (fb :: TYPE repb)), y~(a %q-> (gb :: TYPE repb)) => ToLinearN' ('S k) x y where
```

Reply:
I'll try it.